### ISSUE TITLE
Iceberg connector fails to read metadata

### ISSUE THREAD

**Issue by dmgcodevil** [trinodb/trino#6369]
We've upgraded to 348 since it includes the fix for [iss5825](https://github.com/prestosql/presto/issues/5825)

The new version fails to read metadata for some queries, however, 346 works fine.

````
io.prestosql.spi.PrestoException: Error opening Iceberg split xxxxxxx.parquet (offset=134217728, length=80134319): start index (-4) must not be negative
        at io.prestosql.plugin.iceberg.IcebergPageSourceProvider.createParquetPageSource(IcebergPageSourceProvider.java:439)
        at io.prestosql.plugin.iceberg.IcebergPageSourceProvider.createDataPageSource(IcebergPageSourceProvider.java:208)
        at io.prestosql.plugin.iceberg.IcebergPageSourceProvider.createPageSource(IcebergPageSourceProvider.java:160)
        at io.prestosql.spi.connector.ConnectorPageSourceProvider.createPageSource(ConnectorPageSourceProvider.java:68)
        at io.prestosql.plugin.base.classloader.ClassLoaderSafeConnectorPageSourceProvider.createPageSource(ClassLoaderSafeConnectorPageSourceProvider.java:66)
        at io.prestosql.split.PageSourceManager.createPageSource(PageSourceManager.java:64)
        at io.prestosql.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:254)
        at io.prestosql.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:182)
        at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:319)
        at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
        at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
        at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
        at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
        at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
        at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
        at io.prestosql.operator.WorkProcessorUtils.lambda$processStateMonitor$2(WorkProcessorUtils.java:200)
        at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
        at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
        at io.prestosql.operator.WorkProcessorUtils.lambda$finishWhen$3(WorkProcessorUtils.java:215)
        at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
        at io.prestosql.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:149)
        at io.prestosql.operator.Driver.processInternal(Driver.java:379)
        at io.prestosql.operator.Driver.lambda$processFor$8(Driver.java:283)
        at io.prestosql.operator.Driver.tryWithLock(Driver.java:675)
        at io.prestosql.operator.Driver.processFor(Driver.java:276)
        at io.prestosql.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1076)
        at io.prestosql.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
        at io.prestosql.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:484)
        at io.prestosql.$gen.Presto_348____20201216_222339_2.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IndexOutOfBoundsException: start index (-4) must not be negative
        at io.airlift.slice.Preconditions.checkPositionIndexes(Preconditions.java:81)
        at io.airlift.slice.Slice.checkIndexLength(Slice.java:1302)
        at io.airlift.slice.Slice.slice(Slice.java:870)
        at io.prestosql.parquet.reader.MetadataReader.readFooter(MetadataReader.java:92)
        at io.prestosql.plugin.iceberg.IcebergPageSourceProvider.lambda$createParquetPageSource$9(IcebergPageSourceProvider.java:356)
        at io.prestosql.plugin.hive.authentication.NoHdfsAuthentication.doAs(NoHdfsAuthentication.java:23)
        at io.prestosql.plugin.hive.HdfsEnvironment.doAs(HdfsEnvironment.java:96)
        at io.prestosql.plugin.iceberg.IcebergPageSourceProvider.createParquetPageSource(IcebergPageSourceProvider.java:356)
        ... 31 more
````

cc / @phd3

**Comment by phd3** [comment]
@dmgcodevil thanks for reporting, can you please share the file size seen on disk for the `xxxxx.parquet` file corresponding to the error message? want to verify if this could be related to https://github.com/prestosql/presto/pull/6174.

**Comment by phd3** [comment]
cc @electrum

**Comment by dmgcodevil** [comment]
different sizes, e.g.: 173Mb, 156Mb, 1.55Mb, 43Mb. 

Also, I think the title isn't quite correct because those parquet files are data files, not metadata files...

**Comment by phd3** [comment]
@dmgcodevil After #6174, we start relying on filesize provided by iceberg metadata instead of computing it again through a file system call. 

In this case, the iceberg metadata is likely wrong, and stores the file size as 0. That causes [this line](https://github.com/prestosql/presto/blob/683e1326765dde32d9a44985e2284fb813fd26be/presto-parquet/src/main/java/io/prestosql/parquet/reader/MetadataReader.java#L92) fails because of a negative value of -4. The best thing to do here would be to fix the metadata, since additional FS call for every file is expensive.
 
Could you also share what writer was used for writing the iceberg metadata? That'd be great in helping other users having the same issue.

From presto side, I think we can temporarily provide a session property in the next release to re-enable the stat call in presto, while metadata is being fixed. @electrum does that seem reasonable?

**Comment by dmgcodevil** [comment]
@phd3 we're using spark iceberg writer, the Iceberg version is `0.8.0-incubating`. I realized that presto-iceberg-connector `348` uses Iceberg `0.9.0`,  maybe that's why the metadata written by Spark is missing file sizes.

**Comment by dmgcodevil** [comment]
@phd3 this [line](https://github.com/prestosql/presto/blob/683e1326765dde32d9a44985e2284fb813fd26be/presto-parquet/src/main/java/io/prestosql/parquet/reader/MetadataReader.java#L74) reads parquet metadata, not iceberg metadata. Am I right?

**Comment by dmgcodevil** [comment]
Quick update: I figured out that some iceberg manifest files contain invalid file size info, i.e. parquet file size specified in the manifest file != the actual file size stored in s3. So it's either iceberg-spark calculates or s3 reports the wrong file size

**Comment by phd3** [comment]
> @phd3 this line reads parquet metadata, not iceberg metadata. Am I right?

Yes, as you might've already figured out, it reads parquet metadata. but incorrect Iceberg metadata can cause undesired behavior since Presto relies on it for reading the parquet file.

**Comment by dmgcodevil** [comment]
> > @phd3 this line reads parquet metadata, not iceberg metadata. Am I right?
> 
> Yes, as you might've already figured out, it reads parquet metadata. but incorrect Iceberg metadata can cause undesired behavior since Presto relies on it for reading the parquet file.

There is a problem with Iceberg Parquet writer, I've already reproduced the issue.  Can be a big problem for existing clients...

**Comment by phd3** [comment]
Closed via #6539
